### PR TITLE
FTTD-7 Store Activities in MongoDB

### PR DESCRIPTION
**Pull request by snaanortey**
New updates to the project were to integrate the API to MongoDB by:

-  Creating schema and model definitions for `activity` and `suggestion` 
- Refactoring the handlers from file storage to MongoDB collection

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,65 +1,45 @@
...
+const mongoose = require("mongoose");
+
+// Import random activity route
+const randomActivity = require("./routes/randomActivity");
+
+// Import suggested activity route
+const suggestActivity = require("./routes/suggestActivity");
+
+// Connects server to mongodb compass and atlas
+const connectionstring =
```
Comment (ehimah) [inline index.js]: it doesn't connect to Compass. Compass is an app that runs on your desktop, Atlas is the cloud instance of MongoDB
```suggestion
// Connects the API server to MongoDB Atlas
const connectionstring =
```

File: routes/randomActivity.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+app.get("/random", async function (request, response) {
+  const aggregates = await ActivityModel.aggregate([{ $sample: { size: 1 } }]);
+
+  const activity = {
+    description: aggregates[0].description,
+    references: aggregates[0].references,
+  };
+
+  console.log(activity);
+
```
Comment (ehimah) [inline routes/randomActivity.js]: clean up

File: routes/suggestActivity.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+  // Add new suggested activity to an array
+  // arrayOfSuggestedIdeas.push(suggestedActivity);
+
+  // console.log(arrayOfSuggestedIdeas);
+
+  // Save the array of suggested activities to the suggestions file (suggestions.json)
+  // const data = JSON.stringify(arrayOfSuggestedIdeas, null, 2);
+
+  // await writeFile("suggestions.json", data);
```
Comment (ehimah) [inline routes/suggestActivity.js]: clean up!

File: routes/suggestActivity.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  const isValid = isActivityValid(suggestedActivity);
+
+  //Validate the input from the user and return bad request if the input is invalid
+  if (!isValid) {
+    response.status(400).send("Your suggestion is invalid. Please try again");
+    return;
+  }
+
+  // Suggested activity is saved and passed to databased
+
```
Comment (ehimah) [inline routes/suggestActivity.js]:
```suggestion
  // Suggested activity is saved to the database

```

File: routes/suggestActivity.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+app.post("/suggest", async function (request, response) {
+  // The suggested activity is stored in the body of the http request
+  // const { body } = request;
+  // const suggestedActivity = body;
+
+  // The suggested activity is stored in the mongodb database
+  const suggestedActivity = new suggestionModel(request.body);
+
+  // Calling the function to check that the object passed to the funtion is valid
+  const isValid = isActivityValid(suggestedActivity);
```
Comment (ehimah) [inline routes/suggestActivity.js]: The creation of a MongoDB model object should be done only after we know that the request body is valid
```suggestion
  // Calling the function to check that the object passed to the function is valid
  const isValid = isActivityValid(request.body);
  
    // create an instance of a MongoDB model to be stored to db
  const suggestedActivity = new suggestionModel(request.body);
```

**Review by ehimah — CHANGES_REQUESTED** [review]
fix up

**Comment by snaanortey** [comment]
updated comments, thanks

File: models/activity.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    required: true,
+    trim: true,
+    validate(value) {
+      if (value.length > 250) {
+        throw new Error(
+          "Description is too long. Must be less than 250 characters."
+        );
+      }
+    },
+  },
```
Comment (ehimah) [inline models/activity.js:47]: USE the validation `maxLength` key instead.

File: routes/suggestActivity.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+  // create an instance of a MongoDB to be stroed to db
+  const suggestedActivity = new suggestionModel(request.body);
+
+  // Valid suggested activity is saved to the database
+  try {
+    await suggestedActivity.save();
+    response.send("Suggestion received, thank you!");
+    console.log("try done");
+
```
Comment (ehimah) [inline routes/suggestActivity.js:29]: clean up

File: models/activity.js [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    required: function () {
+      return this.url.length > 0;
+    },
+  },
+  url: {
+    type: String,
+    required: false,
+    validate(value) {
+      return /^(ftp|http|https):\/\/[^ "]+$/.test(value);
+    },
```
Comment (ehimah) [inline models/activity.js:26]: use the match validation key instead.
```suggestion
    match: /^(ftp|http|https):\/\/[^ "]+$/,
```

File: index.js [hunk 9]
```diff
@@ -1,65 +1,45 @@
...
-
-// Import data from suggestions.json so it can be added to arrayOfSuggestedIdeas
-const arrayOfSuggestedIdeas = require("./suggestions.json");
+const mongoose = require("mongoose");
+
+// Import random activity route
+const randomActivity = require("./routes/randomActivity");
+
+// Import suggested activity route
+const suggestActivity = require("./routes/suggestActivity");
```
Comment (ehimah) [inline index.js:10]:
```suggestion
// Import suggest activity route
const suggestActivity = require("./routes/suggestActivity");
```

File: models/activity.js (hunk 6 above)
Comment (snaanortey) [reply models/activity.js:47]: updated

File: routes/suggestActivity.js (hunk 7 above)
Comment (snaanortey) [reply routes/suggestActivity.js:29]: done

File: models/activity.js (hunk 8 above)
Comment (snaanortey) [reply models/activity.js:26]: done

File: index.js (hunk 9 above)
Comment (snaanortey) [reply index.js:10]: done

**Comment by snaanortey** [comment]
update

**Review by ehimah — APPROVED** [review]
Ehimah Approves!

ehimah-very-excited

**Pull request merged**